src/order/grade.lean

```diff
@@ -0,0 +1,334 @@
+/-
+Copyright (c) 2022 Yaël Dillies, Violeta Hernández Palacios. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Yaël Dillies, Violeta Hernández Palacios, Grayson Burton, Vladimir Ivanov
+-/
+import data.nat.interval
+import data.nat.succ_pred
+import order.atoms
+
+/-!
+# Graded orders
+
+This file defines graded orders, also known as ranked orders.
+
+A graded order is an order with a bottom element in which every element has some finite "height",
+that corresponds to how many elements you need to get down by to reach `⊥`.
+
+## Main declarations
+
+* `grade_order`: Graded orders.
+* `grade`: The grade of an element.
+* `grade_max_order`: Graded orders with maximal elements. All maximal elements have the same grade.
+* `max_grade`: The maximum grade in a `grade_max_order`.
+* `order_embedding.grade`: The grade of an element in a linear order as an order embedding.
+
+## Implementation notes
+
+The standard approach is to define graded orders as the bounded orders whose flags (maximal chains)
+all have the same finite length (see Stanley p. 99).
+
+However, this means that graded orders all have a top and that we must reconstruct the grade of an
+element by how many elements we need to go down to reach `⊥`. Unpractical, really.
+
+Instead, we define graded orders by their grade function, without talking about flags yet.
+
+## References
+
+* [Konrad Engel, *Sperner Theory*][engel1997]
+* [Richard Stanley, *Enumerative Combinatorics*][stanley2012]
+-/
+
+open finset nat order_dual
+
+variables {α β : Type*}
+
+/-- A graded order is an order equipped with a grade function which tells how far a given element is
+away from the minimal elements. Precisely, `grade a` is the height of `a` in the Hasse diagram of
+`α`. -/
+class grade_order (α : Type*) [preorder α] :=
+(grade : α → ℕ)
```

Comment:
Is there a reason you prefer to grade by the naturals, and not anything more general? Some authors seem to prefer grading by integers, why do you only allow for this special case?

Reply:
Grading by naturals arises in many places. Typically, `finset.card` grades `finset α`, `id` grades `ℕ`, `dim` grades `affine_subspace`... Grading by `ℤ` is much rarer.

However, if you think that's important, we can make `grade_order` take the codomain as an argument.

Reply:
Grading by naturals seems justified on the grounds that this gives each element of the poset a "dimension," if you're thinking about the preorder as being some extremely abstract complex of polytopes.

How important is it to make this `grade` function be attached data? It seems like you would be better served by a definition where the `grade` for any `preorder` is the `Sup` of lengths of strictly descending chains, then have a property class (`is_grade_order`?) containing the fact that `grade` is a `strict_mono` satisfying `grade_of_covby`. (Some bikeshedding, you could call this the `height` function if it's not taken, but :shrug:)

If you're not doing `#eval`, I don't think you should be worrying about computability. Even if you're worried about computability, you can give computable versions for specific cases separately.

Reply:
No, this is bad because in many cases we have an explicit "canonical" formula:
* `finset.card` for `finset`
* `P.card - 1` for `finpartition`
* `grade x.1 + grade x.2` for `prod`

More crucially, this is needed to define Whitney numbers (see branch `polytopes`) and other properties of the order.

Your motivation seems to be motivated by the fact that `grade_order α` is a subsingleton. That's not a reason to make it noncomputable. Most order classes are subsingletons: `order_top`/`order_bot`/`bounded_order`, `succ_order`/`pred_order`, `lattice`, `complete_lattice` if they were mixins... but it's still important to keep them as is.

Actually, having an instance be a subsingleton seems to be an incentive to keep the data: You can always make it noncomputable from the existence. This is not the case when the data is not a subsingleton but does not contribute anything mathematically useful, as in p-Banach spaces.

Reply:
My motivations are that `Sup` gives a universal canonical formula and that computability in Lean is actually not that meaningful. (Whether something is `noncomputable` is purely a function of whether Lean thinks the VM can compile the definition. It's not checked by the kernel or anything.)

I think the only valid justification for keeping it being data is if the API of `grade_order` is designed to let you rely on the underlying definitions of the `grade` functions that come from instances (i.e., rely on defeq properties). Is this really that useful to you for `finset.card`? I think the only thing you get is that when `s` is some concrete `finset` with known cardinality `n`, that Lean can typecheck `rfl : grade s = n`.

With the other examples you give, like `order_top` and `order_bot`, people tend to rely on the fact that the underlying definition is available without a rewrite. Like for `nat` the fact that `bot` actually is `0` is nice.

If it were a property class, i'd imagine the proof that `finset` is a `grade_order` would start by showing that for all `s`, `grade s = s.card`, then the rest is straightforward. From then on, to use the fact that `grade` is `finset.card` you'd need to rewrite -- but you would have needed to unfold `grade` to do further rewrites anyway, right?

Reply:
A graded order can equivalently be defined as a preorder where all maximal chains have the same length. A grade function can then be assigned based on that. In practice, this condition is much harder to work with and reason about. I believe that this should be a theorem, and not a definition.

Reply:
Further, this theorem does not generalize to infinite graded orders, so that would have to be rethought or removed.

Reply:
I'm not saying that this is what you should do, but this is what I was suggesting:
```lean
import data.list.defs
import data.nat.interval
import data.nat.lattice
import data.nat.succ_pred
import order.atoms

variables {α : Type*} [preorder α]

def down_chains_from (x : α) := {l : list α | l.chain (>) x}

noncomputable
def grade (x : α) : ℕ := Sup (list.length '' down_chains_from x)

class grade_order (α : Type*) [preorder α] : Prop :=
(down_chains_bdd (x : α) : bdd_above (list.length '' down_chains_from x))
```
I'm pretty sure this is equivalent to the definition you have (maybe seeing it this way justifies the notion they've developed of a graded order @b-mehta?), but it has the property that `grade` depends on nothing but the preorder.

Just for further illustration, here are some hastily written low-quality proofs that `nat` and `finset` are a `grade_order` and that `grade n = n` and `grade s = s.card`:
```lean
instance : grade_order ℕ :=
⟨begin
  intros x,
  use x,
  rintro _ ⟨l, h, rfl⟩,
  simp [down_chains_from] at h,
  induction h,
  { simp, },
  { simp,
    linarith, },
end⟩

lemma nat.grade_eq (n : ℕ) : grade n = n :=
begin
  apply le_antisymm,
  { apply cSup_le ((nonempty_down_chains_from _).image _),
    intros m,
    simp [down_chains_from],
    rintros l c rfl,
    induction c,
    { simp, },
    { simp,
      linarith, } },
  { refine le_cSup (grade_order.down_chains_bdd _) _,
    use (list.range n).reverse,
    simp [down_chains_from],
    induction n,
    { simp, },
    { simp [list.range_succ, lt_add_one, *], } },
end

instance (α : Type*) : grade_order (finset α) :=
⟨begin
  intros x,
  use x.card,
  rintro _ ⟨l, c, rfl⟩,
  simp [down_chains_from] at c,
  induction c with h s t c hst hc hl,
  { simp },
  { simp,
    have : s.card > t.card := finset.card_lt_card hst,
    rw nat.add_one_le_iff,
    linarith, },
end⟩

lemma finset.grade_eq {α : Type*} (s : finset α) : grade s = s.card :=
begin
  apply le_antisymm,
  { apply cSup_le ((nonempty_down_chains_from _).image _),
    intros m,
    simp [down_chains_from],
    rintros l c rfl,
    induction c with h s t c hst hc hl,
    { simp },
    { simp,
      linarith [finset.card_lt_card hst], }, },
  { refine le_cSup (grade_order.down_chains_bdd _) _,
    classical,
    refine finset.induction _ _ s,
    { use list.nil,
      simp [down_chains_from] },
    { rintros a s ha ⟨l, c, hl⟩,
      use (s :: l),
      simp [down_chains_from] at c,
      simp [down_chains_from, finset.ssubset_insert ha, c, ha, hl], }, },
end
```

Reply:
> In practice, this condition is much harder to work with and reason about (as mentioned previously, there's usually nice canonical grade functions).

The general theory doesn't get to use the nice canonical grade functions, though.  With the above strategy, there's a truly canonical grade function (the supremum over lengths of descending chains, maximal ones of which are flags), but the tradeoff is that you need to rewrite using `nat.grade_eq` or `finset.grade_eq` if you want to use a more convenient representation of the function.

In private communication, Yael tells me that you want to make use of defeq properties of `grade` (to be able to write term proofs, for example treating `finset.card` and `grade` as being exactly the same), which answers my question from last week where I said this would be justification to include `grade` as data.

> More crucially, this is needed to define Whitney numbers (see branch polytopes) and other properties of the order.

I'm not sure where defeq properties of `grade` come into the definition of Whitney numbers. With the code above, this seems to work:
```lean
class slice_order (α : Type*) [preorder α] :=
(slice [] (n : ℕ) : finset α)
(mem_slice (n : ℕ) (a : α) : a ∈ slice n ↔ grade a = n)

def whitney_number [preorder α] [slice_order α] (n : ℕ) : ℕ :=
(slice_order.slice α n).card
```

Reply:
We're now grading by something more general, and Kyle's universal canonical formula has been sidestepped in the process.